stablehlo/dialect/VhloTypes.td

```diff
@@ -60,9 +63,92 @@ class VHLO_TypeDef<string name,
 // VHLO Type Definitions.
 //===----------------------------------------------------------------------===//
 
-// Token type.
-def VHLO_Token : VHLO_TypeDef<"Token"> {
-  let mnemonic = "token";
+def VHLO_Token : VHLO_TypeDef<"Token", "token">;
+
+//===----------------------------------------------------------------------===//
+// Forked Types
+//===----------------------------------------------------------------------===//
+
+def VHLO_BFloat16V1 : VHLO_TypeDef<"BFloat16V1", "bf16">;
+
+def VHLO_ComplexV1 : VHLO_TypeDef<"ComplexV1", "complex", "::mlir::ComplexType"> {
+  let parameters = (ins "Type":$elementType);
+  let assemblyFormat = "`<` $elementType `>`";
+}
+
+def VHLO_Float16V1 : VHLO_TypeDef<"Float16V1", "f16">;
+
+def VHLO_Float32V1 : VHLO_TypeDef<"Float32V1", "f32">;
+
+def VHLO_Float64V1 : VHLO_TypeDef<"Float64V1","f64">;
+
+def VHLO_IndexV1 : VHLO_TypeDef<"IndexV1", "index">;
+
+def VHLO_IntegerType : VHLO_TypeDef<"IntegerV1", "integer"> {
+  let parameters = (ins "::mlir::IntegerType":$value);
+  let assemblyFormat = "`<` $value `>`";
+  let genVerifyDecl = 1;
+  let extraClassDefinition = [{
+    ::mlir::LogicalResult IntegerV1Type::verify(::llvm::function_ref<InFlightDiagnostic()> emitError,
+                     ::mlir::IntegerType value) {
+      llvm::SmallVector<unsigned> validWidths{4, 8, 16, 32, 64};
+      bool isPred = (value.getWidth() == 1 && value.isSignless());
+      if (!isPred && !llvm::is_contained(validWidths, value.getWidth())) {
+        // return emitError() << "invalid integer width " << value;
+      }
+      //if (!value.isSignless() && !value.isUnsigned()) {
+      //  return emitError() << "invalid integer signedness " << value;
+      //}
+      return success();
+    }
+  }];
+}
+
+def VHLO_RankedTensorV1 : VHLO_TypeDef<"RankedTensorV1", "tensor"> {
+  let parameters = (ins
+    VHLO_Dims:$shape,
+    "::mlir::Type":$elementType,
+    "::mlir::Attribute":$encoding
+  );
+  let assemblyFormat = "`<` custom<TensorShape>($shape) `` $elementType `` custom<Encoding>($encoding) `>`";
+}
+
+def VHLO_TupleV1 : VHLO_TypeDef<"TupleV1", "tuple"> {
+  let parameters = (ins ArrayRefParameter<"::mlir::Type">:$types);
+  let assemblyFormat = "`<` $types `>`";
+}
+
+def VHLO_Double : APFloatParameter<""> {
+  // Used by UniformQuantizedV1 to have APFloat with custom parser
+  let parser = [{
+    [&]() -> FailureOr<llvm::APFloat> {
+      double value;
+      if (failed($_parser.parseFloat(value))) {
+        return failure();
+      }
+      return APFloat(value);
```

Comment:
Why do we need `APFloat` here? The upstream type uses `double`.

Reply:
The resulting tablegen files aren't happy with "double", I couldn't find any TD files for attrs/types that explicitly use the double type either, end up with a build warning on hashing, happy to discuss offline we want to investigate further:

```
/llvm-project/llvm/include/llvm/ADT/Hashing.h:382:10: error: no matching function for call to 'hash_value'
  return hash_value(value);
         ^~~~~~~~~~
/llvm-project/llvm/include/llvm/ADT/Hashing.h:563:63: note: in instantiation of function template specialization 'llvm::hashing::detail::get_hashable_data<double>' requested here
    buffer_ptr = combine_data(length, buffer_ptr, buffer_end, get_hashable_data(arg));
                                                              ^
/llvm-project/llvm/include/llvm/ADT/Hashing.h:566:12: note: in instantiation of function template specialization 'llvm::hashing::detail::hash_combine_recursive_helper::combine<double, long, long, long>' requested here
    return combine(length, buffer_ptr, buffer_end, args...);
           ^
```

Reply:
Looks like UniformQuantizedType gets around this by specifying its own hash function and casting away from double:
```
    unsigned getHashValue() const {
      int64_t scaleBits = llvm::bit_cast<int64_t>(scale);
      return llvm::hash_combine(flags, storageType, expressedType, scaleBits,
                                zeroPoint, storageTypeMin, storageTypeMax);
    }
```

Reply:
Would be nice to get rid of `VHLO_DoubleV1` if we can pull this off.

Reply:
Do you have numeric concerns? I'm not entirely sure of the differences in APFloat-of-double vs double.

This type is also used by FloatAttr so that we can get printing and parsing without needing to implement a custom printer/parser in `VhloOps.cpp` for APFloat types.

Reply:
I think we should use APFloat because it also carries float semantics - something that would be good to preserve. Also, I've just noticed that upstream FloatAttr is using APFloat: https://github.com/llvm/llvm-project/blob/62bd944e42472b7aa01fe6f662d848d76a96b8f8/mlir/include/mlir/IR/BuiltinAttributes.td#L617-L618.

Reply:
This currently uses APFloat, I renamed the type to VHLO_APFloatV1 accordingly.